tephra-core/src/test/java/co/cask/tephra/distributed/PooledClientProviderTest.java

```diff
@@ -108,39 +112,62 @@ private void startClientAndTestPool(Configuration conf) throws InterruptedExcept
     final PooledClientProvider clientProvider = new PooledClientProvider(conf,
       injector.getInstance(DiscoveryServiceClient.class));
 
+    // test simple case of get + return. Note: this also initializes the provider's pool, which
+    // takes about one second (discovery). Doing it before we test the threads makes it so that one
+    // thread doesn't take exceptionally longer than the others.
+    try (CloseableThriftClient closeableThriftClient = clientProvider.getCloseableClient()) {
+      // do nothing with the client
+    }
+
     //Now race to get MAX_CLIENT_COUNT+1 clients, exhausting the pool and requesting 1 more.
     List<Future<Integer>> clientIds = new ArrayList<Future<Integer>>();
-    ExecutorService executor = Executors.newFixedThreadPool(MAX_CLIENT_COUNT + 2);
+    ExecutorService executor = Executors.newFixedThreadPool(MAX_CLIENT_COUNT + 1);
     for (int i = 0; i < MAX_CLIENT_COUNT + 1; i++) {
-      clientIds.add(executor.submit(new RetrieveClient(clientProvider)));
+      clientIds.add(executor.submit(new RetrieveClient(clientProvider, CLIENT_OBTAIN_TIMEOUT / 2)));
     }
 
     Set<Integer> ids = new HashSet<Integer>();
     for (Future<Integer> id : clientIds) {
       ids.add(id.get());
     }
+
+    // now, try it again with, where each thread holds onto the client for twice the client.obtain.timeout value.
+    // one of the threads should throw a TimeOutException, because the other threads dont release their clients
+    // within the configured timeout.
+    try {
+      for (int i = 0; i < MAX_CLIENT_COUNT + 1; i++) {
+        clientIds.add(executor.submit(new RetrieveClient(clientProvider, CLIENT_OBTAIN_TIMEOUT * 2)));
+      }
+      for (Future<Integer> clientId : clientIds) {
+        clientId.get();
+      }
+      Assert.fail(String.format("Expected one thread to not obtain a client within %s milliseconds.",
+                                CLIENT_OBTAIN_TIMEOUT));
+    } catch (ExecutionException expected) {
+      Assert.assertEquals(TimeoutException.class, expected.getCause().getClass());
```

Comment:
Don't we need to assert that the other `MAX_CLIENT_COUNT` clients did not throw any exception?